src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,238 @@
+      previous_most_recent = std::move(*tmp_element);
+    }
+    absl::MutexLock lock(&mu_);
+    if (elements_.empty()) {
+      oldest_enqueued_timestamp_.store(previous_ts, std::memory_order_relaxed);
+    }
+    elements_.push_back(Storage{std::move(previous_most_recent), previous_ts});
+  }
+
+ private:
+  // Attempts to pop from the front of the queue (oldest).
+  // This will return nullopt if the queue is empty, or if other workers
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryLockAndPop(bool front) ABSL_LOCKS_EXCLUDED(mu_) {
+    // Do not block the worker if there are other workers trying to pop
+    // tasks from this queue.
+    if (!mu_.TryLock()) return absl::nullopt;
+    auto mu_cleanup = absl::MakeCleanup([this]() {
+      mu_.AssertHeld();
+      mu_.Unlock();
+    });
+    if (GPR_UNLIKELY(elements_.empty())) {
+      if (most_recent_element_enqueue_timestamp_.load(
+              std::memory_order_relaxed) == kInvalidTimestamp) {
+        return absl::nullopt;
+      }
+      if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+      absl::optional<T> ret = absl::nullopt;
+      if (GPR_LIKELY(most_recent_element_.has_value())) {
+        most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                     std::memory_order_relaxed);
+        ret = std::exchange(most_recent_element_, absl::nullopt);
+      }
+      most_recent_element_lock_.Unlock();
+      return ret;
+    }
+    // the queue has elements, let's pop one and update timestamps
+    Storage ret_s;
+    if (front) {
+      ret_s = std::move(elements_.front());
+      elements_.pop_front();
+    } else {
+      ret_s = std::move(elements_.back());
+      elements_.pop_back();
+    }
+    if (elements_.empty()) {
+      oldest_enqueued_timestamp_.store(kInvalidTimestamp,
+                                       std::memory_order_relaxed);
+    } else if (front) {
+      oldest_enqueued_timestamp_.store(elements_.front().enqueued(),
+                                       std::memory_order_relaxed);
+    }
+    return ret_s.TakeElement();
+  }
+
+  // Attempts to pop from the back of the queue (most recent).
+  // This will return nullopt if the queue is empty, or if other workers
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryPopMostRecentElement() {
+    if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+    if (GPR_UNLIKELY(!most_recent_element_.has_value())) {
+      most_recent_element_lock_.Unlock();
+      return absl::nullopt;
+    }
+    most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                 std::memory_order_relaxed);
+    absl::optional<T> tmp = std::exchange(most_recent_element_, absl::nullopt);
+    most_recent_element_lock_.Unlock();
+    return tmp;
+  }
+
+  // The managed items in the queue
+  std::deque<Storage> elements_ ABSL_GUARDED_BY(mu_);
```

Comment:
Note that `std::deque` has a minimum bucket size of 4KB, meaning we're likely to be consuming that much extra per thread - is it warranted?

Reply:
Internally, I have implemented a linked ring-buffer.

I suspect we won't use this class internally and use the executors we have, right?

Reply:
Internally in the data center I expect we won't use this class.
We will for any non data center apps (mobile, other small devices).

Reply:
Most of the 4k would be unused in our current use cases, so I can see about a ring buffer implementation. From what I've seen so far though, I'd expect under 10 threads in the pool for the current workloads, so < 50Kb added memory with the global queue included. That's well within the "medium bloat" labeling.